src/v/storage/segment_deduplication_utils.cc

```diff
@@ -196,7 +204,7 @@ ss::future<index_state> deduplicate_segment(
               return ss::make_ready_future<bool>(false);
           }
 
-          return should_keep(map, b, r);
+          return should_keep(map, b, r, may_have_tombstone_records);
```

Comment:
It was written this way because `should_keep()` returns a `ss::future<bool>`. We would have to make this a lambda-coroutine to write it in the cleaner way you posted.

Reply:
Altered code with a continuation.